### ISSUE TITLE
Open Nomad links in same Window

### ISSUE THREAD

**Issue by piotrzarzycki21** [Moonshine-IDE/Super.Human.Portal#53]
From HCL:

>Greetings Justin and the Prominic team,
> 
>Thank you for meeting with us to discuss the Nomad team's progress regarding a number of issues. I have attached the slide deck shared in the 20220222 checkpoint meeting detailing the launch handler and link capturing feature experiments in Chrome and Nomad Web. Since we last spoke, the Chromium team has communicated that this experimental feature will not be available in Chrome release 123 due to issues they're seeing with the current implementation. We will continue to monitor and test their changes accordingly in order to improve the link handling experience in Nomad.
> 
>On slides 3-5, I've added a number of scenarios and the observed behaviors that an end user may experience. Each of these are under the condition that both the Chrome experiment and Nomad experiment flags are enabled.
> 
>We also discussed an alternative solution that may suit Prominic's needs by using service worker messaging to open links in an existing Nomad browser tab or PWA window. Slide 6 is an explanation of how Nomad Web is using this messaging API to trigger the opening of links in the main Nomad window if the browser has opened an intermediate secondary window. On slides 7-9, I have included an example web page, scopedlinks.html, that leverages that functionality by overriding the browser's default behavior of clicking on a link to use the navigator.serviceWorker.controller.postMessage() API instead. You may edit and test this file with links appropriate to your deployment. Paste this file in the nomad-files/ directory with the rest of the Nomad Web client files. This will ensure that the postMessage() calls are messaging Nomad's service worker. Note that this is not using launch handler or link capturing features. The experiments do not need to be Enabled to test this page.
> 
>Thank you for your time to continue to discuss these issues as we work toward a solution. Please feel free to reach out if you have any further questions.

[scopedlinks.html.zip](https://github.com/Moonshine-IDE/Super.Human.Portal/files/14608311/scopedlinks.html.zip)

Summary from @piotrzarzycki21

>Here is the context of our short chat today. They provided example in html file, I have added that code to Super.Human.Portal in Browse My Server - select some application and click "Open in Nomad" their code will be executed. Lates build should have my changes.
>If you open Web Developer tools and go to file MediatorBrowseMyServer jump to method: onOpenNomadWeb - place there debugging break point and you will be able see that ServiceWorker is in pending state and controller is null.
>
>I'm wondering based on their description below whether we need some special deployment for Super.Human.Portal to have it working. Maybe you will have some thoughts on that - if not let me know and I will respond to this email asking for assistance.

UPDATE:  Branch:  https://github.com/Moonshine-IDE/Super.Human.Portal/tree/features/issue_53_nomad_links

**Comment by JoelProminic** [comment]
This is experimental functionality, so I'd prefer to keep it in a branch so that it doesn't interfere with the 0.1.9 release (which I have been trying to return to).

**Comment by piotrzarzycki21** [comment]
You are right! I'm sorry about that - I will revert my changes in master !

**Comment by JoelProminic** [comment]
My test Domino+Nomad instances are in a bad state at the moment, so I may run out of time to test this today.

Were you able to get the example HTML file to work?  By my understanding, the process is:
1. Open Nomad once with the exp_link_capturing option:  https://nomadweb.demo.startcloud.com/nomad/?exp_link_capturing=true 
2. Upload the HTML file to the Domino server (i.e. `/local/notesdata/domino/html/scopedlinks.html`)
3. Open the test HTML file. Note that the subdomain is different:  https://domino.demo.startcloud.com/scopedlinks.html
4. Click on one of the links. 

I originally tested with Chrome, but this failed with an error on step 1 (even when I enabled the suggested option).
![image](https://github.com/Moonshine-IDE/Super.Human.Portal/assets/24532731/f11eac4c-4471-4c91-a73e-dde529598a8c)

I got past this on Firefox, but it triggered the the "no serrvice worker registered" case.

My theory is that we would need to be on the same subdomain to properly trigger this.  I'm looking for a way to confirm this.

**Comment by piotrzarzycki21** [comment]
> Were you able to get the example HTML file to work?

No I did not, I don't know how to upload stuff to server.

**Comment by piotrzarzycki21** [comment]
> I originally tested with Chrome, but this failed with an error on step 1 (even when I enabled the suggested option).

It means that your ServiceWorker is probably in the same state is mine - pending, nothing change, but in order to confirm that it would be good to look into console debugger.

**Comment by piotrzarzycki21** [comment]
I have moved code in SHP to [branch](https://github.com/Moonshine-IDE/Super.Human.Portal/tree/features/issue_53_nomad_links), you can grab build [here](https://github.com/Moonshine-IDE/Super.Human.Portal/actions/runs/8294267430)

**Comment by piotrzarzycki21** [comment]
I would like to respond them next week with our tests results. What if we need to be under same subdomain to trigger that - would it change significantly our SHP effort ?

**Comment by JoelProminic** [comment]
> > Were you able to get the example HTML file to work?
> 
> No I did not, I don't know how to upload stuff to server.

@piotrzarzycki21, I was talking with @rat-moonshine about deploying files for another project.  Some options:
- Add it to the build directory for a Royale project and deploy it from Moonshine.   The backend for this is generic, so we may make a generic Moonshine feature for this soon.
- Manully deploy the file with the Vagrant REST API.   This then needs to be copied to a directory under `/local/notesdata/domino/html` and assigned to `domino` (`sudo chown domino:domino <file>`)
- Use the vagrant-scp plugin (and move and update permissions as above):  https://github.com/invernizzi/vagrant-scp 
- Use `vagrant ssh` to connect to the VM, run `sudo su - domino` and paste the contents of the file into your preferred command-line editor. This is the solution I used.

**Comment by piotrzarzycki21** [comment]
> * Add it to the build directory for a Royale project and deploy it from Moonshine.   The backend for this is generic, so we may make a generic Moonshine feature for this soon.

@JoelProminic I think I sort of did that. Because solution is implemented in Super.Human.Portal - I'm deploying this from Moonshine and it's on my server.

**Comment by piotrzarzycki21** [comment]
> * Use the vagrant-scp plugin (and move and update permissions as above):  https://github.com/invernizzi/vagrant-scp

I've tried this but I'm getting permission denied. I think I don't understand something here correctly:
```

piotrzarzycki@Piotrs-MacBook-Pro ~ % vagrant global-status
id       name                         provider   state    directory                                                                                       
----------------------------------------------------------------------------------------------------------------------------------------------------------
ec79cde  default                      virtualbox running  /Users/piotrzarzycki/Dev/machines                                                               
423cb4b  3785--mars.startcloud.com    virtualbox poweroff /Users/piotrzarzycki/Library/Application Support/SuperHumanInstallerDev/servers/demo-tasks/3785 
96031e2  2405--jupiter.startcloud.com virtualbox running  /Users/piotrzarzycki/Library/Application Support/SuperHumanInstallerDev/servers/demo-tasks/2405 
 
The above shows information about all known Vagrant environments
on this machine. This data is cached and may not be completely
up-to-date (use "vagrant global-status --prune" to prune invalid
entries). To interact with any of the machines, you can go to that
directory and run Vagrant, or you can use the ID directly with
Vagrant commands from any directory. For example:
"vagrant destroy 1a2b3c4d"
piotrzarzycki@Piotrs-MacBook-Pro ~ % vagrant scp /Users/piotrzarzycki/Dev/prominic/scopedlinks.html 96031e2:/local/notesdata/domino/html/scopedlinks.html
Warning: Permanently added '[127.0.0.1]:2222' (ED25519) to the list of known hosts.
scp: dest open "/local/notesdata/domino/html/scopedlinks.html": Permission denied
scp: failed to upload file /Users/piotrzarzycki/Dev/prominic/scopedlinks.html to /local/notesdata/domino/html/scopedlinks.html
piotrzarzycki@Piotrs-MacBook-Pro ~ % vagrant scp /Users/piotrzarzycki/Dev/prominic/scopedlinks.html 96031e2:/local/notesdata/domino/html/                
Warning: Permanently added '[127.0.0.1]:2222' (ED25519) to the list of known hosts.
scp: dest open "/local/notesdata/domino/html/scopedlinks.html": Permission denied
scp: failed to upload file /Users/piotrzarzycki/Dev/prominic/scopedlinks.html to /local/notesdata/domino/html/
piotrzarzycki@Piotrs-MacBook-Pro ~ % 
```

**Comment by rat-moonshine** [comment]
Why don't you tried with sudo; until someone suggest you with sudo su?

**Comment by piotrzarzycki21** [comment]
> Why don't you tried with sudo; until someone suggest you with sudo su?

How do I do that ? Could you write step by step what should I do ? Should I open terminal from Super.Human.Installer and type what @JoelProminic propose ? What command ?

**Comment by rat-moonshine** [comment]
Looks like following command gives you permission trouble, try run this in following way:

`sudo vagrant scp /Users/piotrzarzycki/Dev/prominic/scopedlinks.html 96031e2:/local/notesdata/domino/html/scopedlinks.html`

You shall be ask to enter password.

**Comment by piotrzarzycki21** [comment]
> Looks like following command gives you permission trouble, try run this in following way:
> 
> `sudo vagrant scp /Users/piotrzarzycki/Dev/prominic/scopedlinks.html 96031e2:/local/notesdata/domino/html/scopedlinks.html`
> 
> You shall be ask to enter password.

@JoelProminic do you know what's the password to machines?

**Comment by JoelProminic** [comment]
I would do it like this:

    # Use the button in SHI to open the Finder and Terminal window - I think this may bypass the password issue
    # I would copy files to the home directory to avoid permissions issues
    vagrant status      # to get the VM name
    vagrant scp scopedlinks.html 4882--demo.STARTcloud.com:~/
    # Then copy the file to the correct position and permissions
    vagrant ssh
    sudo cp scopedlinks.html /local/notesdata/domino/html/scopedlinks.html
    sudo chown domino:domino /local/notesdata/domino/html/scopedlinks.html

Also note that for a single text file like this, you can do something like:
    
    sudo so - domino
    bash
    cd domino/html
    # replace vi with an editor you are comfortable with
    vi scopedlinks.html
    # Paste the contents of the file and save

**Comment by JoelProminic** [comment]
After more discussion with HCL, I found that we need to create the files in `/opt/hcl/domino/notes/latest/linux/nomad-files`.  For example:

    sudo su -
    cd /opt/hcl/domino/notes/latest/linux/nomad-files
    vi scopedlinks.html
    # Open file at https://nomadweb.%server%/nomad/scopedlinks.html

To put Super.Human.Portal in the directory:

    sudo su -
    cd /opt/hcl/domino/notes/latest/linux/nomad-files/
    ln -s /local/notesdata/domino/html/Super.Human.Portal Super.Human.Portal
    # open with:  https://nomadweb.%server%/nomad/Super.Human.Portal/js-release/index.html

However, this would have problems with the agent URLs and CORS.

As an alternative, here is the haproxy solution we discussed with @MarkProminic today:
- Create the file at `/local/notesdata/domino/html/nomad/scopedlinks.html`
- Update `/etc/haproxy/haproxy.conf` 
    - Add an ACL entry (host_nomadweb_scopedlinks) after the host_nomadweb entry
    - Add a corresponding backend **before** the host_nomadweb entry

```
        ## Host ACL Definitions
        ...
                acl host_nomadweb             hdr(host)          -i nomadweb.demo.STARTcloud.com
                acl host_nomadweb_scopedlinks path_beg          /nomad/scopedlinks.html

        ...
        ## Application backends
        ...
                use_backend                     domino               if host_nomadweb host_nomadweb_scopedlinks
                use_backend                     nomadweb             if host_nomadweb
```

**Comment by JoelProminic** [comment]
From our discussion today, the plan is to have only a "nomadhelper" file inside nomad-files, and have that open the link:
1. Click a link in Super.Human.Portal (under domino.%server%)
2. Call https://nomadweb.%server%/nomad/nomadhelper.html?link=%link%
3. nomadhelper.html opens the link with service worker
4. nomadhelper.html closes the tab automatically after some time or confirming the action.

**Comment by piotrzarzycki21** [comment]
@JoelProminic Issue which you have seen during meeting that only part of Super.Human.Portal was displayed - was real bug. Some of the links to used libraries - DevEpress were blocked for some reason. I changed them and now application is being displayed correctly.

**Comment by piotrzarzycki21** [comment]
Deployed SHP to nomad folders of course won't work due to agents different domain. I just tried to put in nomad folders small js file nomadhelper.js. Loading thins in Royale application ends up with cors issue:

The resource at “https://nomadweb.venus.startcloud.com/nomad/nomadhelper.js” was blocked due to its Cross-Origin-Resource-Policy header (or lack thereof). See https://developer.mozilla.org/docs/Web/HTTP/Cross-Origin_Resource_Policy_(CORP)# index.html

**Comment by JoelProminic** [comment]
Ah, I thought we already had spent time on a CORS workaround for this, but I think I'm confusing this with [another issue](https://github.com/Moonshine-IDE/moonshine-vagrant-mxhx-demo/issues/1)

I think the solution for this is to add the CORS headers for nomadweb.%server% like this:

Run Vagrant SSH and Connect as root: `sudo su -`

Edit `/etc/haproxy/haproxy.cfg` with the command-line editor of your choice.  Add these lines after `backend nomadweb` (note that you need to replace `%yourserver%`):

        http-after-response set-header Access-Control-Allow-Origin "https://domino.%yourserver%"
        http-after-response set-header Access-Control-Allow-Headers "*"
        http-after-response set-header Access-Control-Max-Age "31536000"


Update haproxy with the changes like this:

    @ Verify the file
    haproxy -c -f /etc/haproxy/haproxy.cfg
    # Restart the service
    service haproxy restart
    

Then, open the HTML file you deployed (example:  https://nomadweb.nomad-1010.nomadtest.com/nomad/scopedlinks.html) in a browser and check for the CORS headers.

You can then try this from Super.Human.Portal.  Watch for the preflight issue I reported [here](https://github.com/Moonshine-IDE/moonshine-vagrant-mxhx-demo/issues/1#issuecomment-2024126661) - I'll update the above instructions if I work out  a fix for this.

**Comment by rat-moonshine** [comment]
Without knowing much of the original issue here, but attracted to `HAProxy.cfg` file discussion - I went to comment-off the CORS lines in `haproxy.cfg` and test the MXHX application - but in browser console continuing to have an error:

`POST https://restapi.prominic.domino.com/file/upload?rand=0.26464644545 net::ERR_CERT_AUTHORITY_INVALID`

**Comment by piotrzarzycki21** [comment]
> UPDATE: We received alternative CORS instructions from HCL. I have updated these for our servers:
> 
> ```
> sudo su - domino
> bash
> # edit nomad-config.yml with the command-line editor of your choice.  This file didn't exist already for me
> # open the domino console
> domino console
> # run these commands
> restart task nomad
> close
> ```

@JoelProminic going with that direction means I'm being asked for a password again and I still don't know what's the password to this domino instance.

**Comment by piotrzarzycki21** [comment]
> Example of `nomad-config.yml`. Update the value like this: `https://domino.%yourserver%`
> 
> ```
> httpHeaders:
>   Access-Control-Allow-Origin: https://domino.nomad-1010.nomadtest.com
> ```

By going with suggestion above do I create/modify nomad-config.yml under following location /local/notesdata/nomad-config.yml ?

**Comment by JoelProminic** [comment]
If I remember right `su - domino` requires a password, but `sudo su - domino` should not, at least if the current user is `startcloud`.   I'm not sure if the `domino` user even has a real password defined - I have never needed to use this.

Yes, the file to create/update is `/local/notesdata/nomad-config.yml`.   `/local/notesdata/` is the home directory for the `domino` user, but in retrospect I should have used the absolute path.

**Comment by JoelProminic** [comment]
We ended up needing a fix for preflight for this issue as well.   I added some instructions for this [here](https://github.com/Moonshine-IDE/moonshine-vagrant-mxhx-demo/issues/1#issuecomment-2033035222).

However, we are still getting an error (`ns_error_dom_corp_failed`) when trying to load the nomadweb file as a [JavaScript library](https://github.com/Moonshine-IDE/Super.Human.Portal/blob/8a46a2e3415e94ee8878ea6d863c63f1f031e655/Super.Human.Portal_Royale/src/resources/jewel-index-template.html#L34).

**Comment by piotrzarzycki21** [comment]
I think we should take step back from our investigation. After some discussion with Adrian we are thinking that we need to run nomadhelper.js from server not from client. All our previous attempts (downloading with fetch, load javascript function, requesting trough [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest), dynamically importing) resulted in the code being called in the client not from the server. 

We think we need to force Nomad to run `nomadhelper.js` as a part of the server. 

We think we should respond to Daniel with fillowing message:
- We would like to describe function which shold be called from client (source code)
- Function should be running on Nomad server, NOT client
- When user (Super.Human.Portal) call function on the server trough XMLHttpRequest with parameters - server parse params and calls proper logic with passed parameters.

Pseudo code:

```
function openLink(req, res) => {
  if (req.url '/openLink') {
  	const hrefUrl = new URL(req.link);
    if (navigator?.serviceWorker?.controller) {
      navigator.serviceWorker.controller.postMessage({
        type: 'openNotesUri',
        payload: {
          notesUri: decodeURIComponent(
            hrefUrl.hash.substring(
              hrefUrl.hash.indexOf(HASH_PREFIX) + HASH_PREFIX.length
            ),
          ),
        },
      });
    } else {
      console.error('no service worker registered');
    }

    res.writeHead(200, { 'Content-Type': 'text/plain' });
    res.end('some');
  }
```

  Do you agree with this appraoch ?

**Comment by JoelProminic** [comment]
Here were the solutions proposed by @JustinProminic:


>1) My main suggestion was to try to use a GET request on the shim file to tell it what to do and avoid CORS.
>
>2) The next idea is to just put ALL of Super.Human.Portal under the nomad URL if ew have to with redirects like what Mark showed last week. This should be possible by using HAProxy and the rewrite rules if we have everything under S.H.P. matching a URL such as /SHP/index.html Then you can make SHP load off of nomadweb and work, in theory, from the same base URL.
>
>3) A 3rd idea I have is to put a frameset around the SHP so that part of the page is SHP and part of it is the frame that contains loadjavascript.as and then you are able to call loadjavascript.as from the GET parameter and still avoid CORS.


> Ok but 2 is actually the least favorite option. 1 or 3 would be better. But 2 is worth a try. 

We have been working towards option 1 over the last few days, but when we were working on this yesterday we switched to trying to load the file from Nomad as a library, and then making a call using this.  Unfortunately, I think this will probably still run as `domino.` - did you finish a test for this?

For 1, we would want to make a call to an HTML file on `nomadweb.`.  This HTML file should contain the JavaScript code from the HCL example, as well as JavasScript code to read the query-string parameter and call the method to open the link.  However, I don't think this file will be executed you just read the URL inside Royale - I think this would just return the raw HTML/JavaScript instead.  Is there another way to evaluate this HTML in the local browser window?

I think 3 is promising.  This would use the same HTML file as 1, but it would be opened in a "hidden" iframe instead.  The file would then execute in that iframe.  Note that there may be additional headers required to allow nomad to appear in iframes.

I'm not sure your proposed solution makes sense.  This JavaScript needs to run on the client side in order to detect the local service working and update the local tab.  Even if we have the Nomad server evaluates the GET request for us, it will still need a local environment to run.  We could use iframes like 3, but then this just sounds like an unnecessary complication for 3.

**Comment by piotrzarzycki21** [comment]
> We ended up needing a fix for preflight for this issue as well. I added some instructions for this [here](https://github.com/Moonshine-IDE/moonshine-vagrant-mxhx-demo/issues/1#issuecomment-2033035222).
> 
> However, we are still getting an error (`ns_error_dom_corp_failed`) when trying to load the nomadweb file as a [JavaScript library](https://github.com/Moonshine-IDE/Super.Human.Portal/blob/8a46a2e3415e94ee8878ea6d863c63f1f031e655/Super.Human.Portal_Royale/src/resources/jewel-index-template.html#L34).
> 
> UPDATE: I got [some documentation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cross-Origin_Resource_Policy) for this policy from Firefox. This is a related policy called CORP (there is also CORB). To resolve this, we need to change Cross-Origin-Resource-Policy. I found that this was set to `same-origin` by default, and updating it in `nomad-config.yml` did not change this header.
> 
> I ended up fixing this in `/etc/haproxy/haproxy.cfg`. Add this line to `backend nomadweb` and restart the service
> 
> ```
>         http-after-response set-header Cross-Origin-Resource-Policy "same-site"
> ```
> 
> With this change, I was able to load `nomadhelper.js`. UPDATE 2: I needed to change `Console.log` to `console.log`, and then the debug message worked. You can find this code in the Firefox debugger if needed.

I'm sorry for missing this yesterday. Yes this resolves issue with CORS on my sight. I'm starting to play with idea of loading html website in our app and run nomad Service Worker code.

**Comment by piotrzarzycki21** [comment]
@JoelProminic Something has started to work! I went trough deploying [nomadhelper.html](https://github.com/Moonshine-IDE/Super.Human.Portal/blob/features/issue_53_nomad_links/Super.Human.Portal_Royale/src/resources/nomadhelper.html) and [loading it into our application](https://github.com/Moonshine-IDE/Super.Human.Portal/blob/89799d8b9a86d7ef31b4b155679d6b27cf946aab/Super.Human.Portal_Royale/src/net/apacheRoyaleTemplatedApp/controller/CommandLaunchNomadLink.as#L37). I've decided instead using iFrame to use [JQuery.load](https://api.jquery.com/load/) function. As you can see in code url to nomadhelper i hardcoded - I'm not sure how we should provide that url to our application - if you wanted to test it on your end with different deployment you need to change url in CommandLaunchNomadLink and build Super.Human.Portal yourself.

**Comment by piotrzarzycki21** [comment]
Functionality is working for now only i **Browse My Server** -> select application and use button **Open in Nomad**. Of course you will have to first copy nomadhelper.html

`sudo cp /local/notesdata/domino/html/Super.Human.Portal_Royale/js-debug/resources/nomadhelper.html /opt/hcl/domino/notes/latest/linux/nomad-files`

**Comment by piotrzarzycki21** [comment]
@JoelProminic It looks like yesterday I was too optimistic with above solution - it turns out it didn't work. I've been spending with @Aszusz whole day on trying different approaches. We ended up using iFrame and this is working. I have pushed my code. 

Here is what we had to do:
1. In /local/notesdata/nomad-config.yml you have to add 

`X-Frame-Options: ALLOW-FROM https://domino.venus.startcloud.com`

Once we added that we gain working `window.onload` in nomadhelper.html, but we have started to experience in console log following error in Firefox 124.0.2 (64-bit)

`Invalid X-Frame-Options header was found when loading “https://nomadweb.venus.startcloud.com/nomad/nomadhelper.html?link=https%3A%2F%2Fnomadweb.venus.startcloud.com%2Fnomad%2F%23%2Fnotes%3A%2F%2Fvenus.startcloud.com%2FSuperHumanPortal.nsf”: “invalid” is not a valid directive. nomadhelper.html`

Despite above error we had access to nomad ServiceWorker.

2. If you wanted to use your custom server url you can change it [here](https://github.com/Moonshine-IDE/Super.Human.Portal/blob/8d6fd2c6f96539630ce53e377f11e44e9a3e64fc/Super.Human.Portal_Royale/src/net/apacheRoyaleTemplatedApp/controller/CommandLaunchNomadLink.as#L17)

3. This part is still relevant https://github.com/Moonshine-IDE/Super.Human.Portal/issues/53#issuecomment-2037403402

Please grab latest build and try to test.

**Comment by JoelProminic** [comment]
Okay, I'll try this when I have a chance.   I had some trouble with the testing yesterday, but I ran into some problem (probably not all related to the above change).

One problem that wasted a bunch of my time was that Nomad was very slow for me.   Have you been noticing similar problems on your test VM?  I'm worried that the haproxy or Nomad configuration changes could have had side-effects on the server performance, but I haven't been able to track down the problem yet.

**Comment by piotrzarzycki21** [comment]
> One problem that wasted a bunch of my time was that Nomad was very slow for me. Have you been noticing similar problems on your test VM?

Nope I'm not experiencing slowliness in that matter. Does version of Nomad have anything with that maybe ? I'm using 1.0.10.

**Comment by JoelProminic** [comment]
I was able to resolve the slow response by restarting the browser (macOS Firefox).  I'm not sure why it got into a bad state, but I noticed that I didn't have the same problem when testing with wget/curl or a different browser.

While testing, I am seeing this error on the console:

> Invalid X-Frame-Options header was found when loading “https://nomadweb.nomad-1010.nomadtest.com/nomad/nomadhelper.html?link=https%3A%2F%2Fnomad.dmytro.cloud%3A9443%2Fnomad%2F%23%2Fnotes%3A%2F%2Fnomad-1010.nomadtest.com%2Fnames.nsf”: “invalid” is not a valid directive. nomadhelper.html

However, the database is still opening in the Nomad tab with the above error, so it doesn't seem to be a blocker.

Doing some research, I see that [ALLOW_FROM is obsolete](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options?utm_source=mozilla&utm_medium=firefox-console-errors&utm_campaign=default).  I tried using this instead:

  Content-Security-Policy: frame-ancestors https://domino.nomad-1010.nomadtest.com;

This worked, but I am still getting some warnings.  I don't see `x-frame-options` set anywhere in nomad-config.yml or haproxy.cfg after my change.

>Content-Security-Policy: Ignoring ‘x-frame-options’ because of ‘frame-ancestors’ directive.


Testing some more cases I see:

Test Case | Behavior
-- | --
Nomad is open and logged in | Database opens in existing Nomad tab
Nomad is open but not logged in | After user authenticates, database opens in existing Nomad tab
Nomad is not open | Nothing happens

To support the last case, I think we should open the provided link in a new tab if the service worker is not defined.  I did a quick test with this, but I ran into trouble with the popup blocker:

```
		} else {
			console.error('no service worker registered');
			window.open(link, '_blank'); // open in new tab unless configured otherwise by the browser
		}
```

**Comment by piotrzarzycki21** [comment]
@JoelProminic I have added changes in other places where "**Open in Nomad**" button is displayed. It's in **Installed Apps** and **Bookmarks**. I have as well cleaned up a bit code of nomadhelper.html - deploy that one as well if you wanted to test latest build.

**Comment by JoelProminic** [comment]
Remaining steps:
- @MarkProminic or I will [update the provisioners](https://github.com/STARTcloud/hcl_roles/issues/6) with the required haproxy and nomad-files modifications
- However, not all servers will support this, so I'd like to continue supporting the old method of opening Nomad links.  My plan is to add a new property to [ConfigRead](https://github.com/Moonshine-IDE/Super.Human.Portal/blob/db5e0c084161e0af99aa94f06c766012990f4e9c/Super.Human.Portal_Agents/src/main/java/config/ConfigRead.java#L49-L51) like `nomad_helper_url`
    - If this is empty, then use the original logic
    - if this is not empty, then use this as the URL for nomadhelper.html.  This should also avoid needing to compute the base URL
    - Alternatively, this can handled in the agent logic only, similar to how we use [nomad_base_url](https://github.com/Moonshine-IDE/Super.Human.Portal/blob/db5e0c084161e0af99aa94f06c766012990f4e9c/Super.Human.Portal_Agents/src/main/java/genesis/LinkProcessor.java#L159-L169) to [generate the current nomad URL](https://github.com/Moonshine-IDE/Super.Human.Portal/blob/db5e0c084161e0af99aa94f06c766012990f4e9c/Super.Human.Portal_Agents/src/main/java/genesis/LinkProcessor.java#L159-L169)
- Handle the case when Nomad is not already open ([more info](https://github.com/Moonshine-IDE/Super.Human.Portal/issues/53#issuecomment-2040575108))

**Comment by JoelProminic** [comment]
I added the new parameter:  `nomad_helper_url`

Configuration | Agent Value | Test | Notes
--|--|--|--
No config doc | `''` | Confirmed | Open the nomadURL links in a new tab, using the the old logic.  This will be the default for fresh Super.Human.Portal installations
Config doc empty | `''` |  | Same as above
Other value | return value exactly |  | This should be treated as the base URL for nomadhelper.html.  Assume that haproxy has been configured correctly if this is set


I also created [example JSON](https://github.com/Moonshine-IDE/Super.Human.Portal/tree/features/issue_53_nomad_links/Super.Human.Portal_Agents/data) to update the configuration documents with Genesis.

Document | create_ | update_ |  Test
-- | -- | -- | --
No config doc | Create document | Nothing to do | Confirmed
Config doc empty | BUG:  Create duplicate document | Update document |  Update not working currently
Other value | BUG:  Create duplicate document | Update document | Update not working currently
Same value | BUG:  Create duplicate document | Nothing to do | Update not working currently

If I can get the `update_` script working, I'll plan to update it to:
1. Change the Form for the existing document, if any.  This should hide it from the Configuration view.
2. Create a new document.

**Comment by piotrzarzycki21** [comment]
> I added the new parameter: `nomad_helper_url`
> 
> Configuration	Agent Value	Test	Notes
> No config doc	`''`	Confirmed	Open the nomadURL links in a new tab, using the the old logic. This will be the default for fresh Super.Human.Portal installations
> Config doc empty	`''`		Same as above
> Other value	return value exactly		This should be treated as the base URL for nomadhelper.html. Assume that haproxy has been configured correctly if this is set
> I also created [example JSON](https://github.com/Moonshine-IDE/Super.Human.Portal/tree/features/issue_53_nomad_links/Super.Human.Portal_Agents/data) to update the configuration documents with Genesis.
> 
> Document	create_	update_	Test
> No config doc	Create document	Nothing to do	Confirmed
> Config doc empty	BUG: Create duplicate document	Update document	Update not working currently
> Other value	BUG: Create duplicate document	Update document	Update not working currently
> Same value	BUG: Create duplicate document	Nothing to do	Update not working currently
> If I can get the `update_` script working, I'll plan to update it to:
> 
> 1. Change the Form for the existing document, if any.  This should hide it from the Configuration view.
> 2. Create a new document.

@JoelProminic I have pushed changes which using new parameter. If it's empty I'm using old behaviour to open url in Nomad.

**Comment by piotrzarzycki21** [comment]
@JoelProminic how can I reproduce your environment setup?

**Comment by piotrzarzycki21** [comment]
> In my tests today, I was not able to trigger the nomadhelper logic. I confirmed that nomad_helper_url was set in the agent, and that this was not a caching issue.
> 
> Test	Expected	Result
> nomad_helper_url is empty	Open link in new tab	Confirmed
> nomad_helper_url is populated	Open link in existing tab	BUG: opens using old logic
> nomad_helper_url is invalid	Open link in existing tab (but this will fail)	BUG: opens using old logic
> Unfortunately, this means I could not test the links in the [other interface locations](https://github.com/Moonshine-IDE/Super.Human.Portal/issues/53#issuecomment-2045408122) or the case when [Nomad is not open](https://github.com/Moonshine-IDE/Super.Human.Portal/issues/53#issuecomment-2040575108).

@JoelProminic ignore please my previous question. I think I have fixed at least second issue.

**Comment by JoelProminic** [comment]
I retested this, and it is working like I specified in my earlier comment.  I also confirmed the other interfaces were updated.

Test | Expected | Browse My Server | Bookmarks | Installed Apps 
-- | -- | -- | -- | --
nomad_helper_url is empty | Open link in new tab | Confirmed | Confirmed | Confirmed
nomad_helper_url is populated | Open link in existing tab | Confirmed | Confirmed | Confirmed
nomad_helper_url is invalid | Open link in existing tab (but this will fail) | Confirmed | Confirmed | Confirmeed.

However, I see that when I click the unexpanded button and the default action is Nomad, it uses the old Nomad logic rather than the new logic, even if nomad_helper_url is set.  I had previously only tested this with the Open in Nomad button.

For the "invalid" case, I see that the URL error is detected in the developer tools.  Can we detect this error with nomadhelper.html and open the nmoadURL link in a new tab instead?

Similarly, the case where Nomad is not open already is also a big problem for this updated design.   I remember that the state of the service worker was ambiguous in this case.   If we can detect the problem, we could open the nomadURL in a new tab instead.

**Comment by piotrzarzycki21** [comment]
> However, I see that when I click the unexpanded button and the default action is Nomad, it uses the old Nomad logic rather than the new logic, even if nomad_helper_url is set. I had previously only tested this with the Open in Nomad button.

This bug has been addressed, please deploy latest build.

**Comment by piotrzarzycki21** [comment]
> For the "invalid" case, I see that the URL error is detected in the developer tools. Can we detect this error with nomadhelper.html and open the nmoadURL link in a new tab instead?
> 
> Similarly, the case where Nomad is not open already is also a big problem for this updated design. I remember that the state of the service worker was ambiguous in this case. If we can detect the problem, we could open the nomadURL in a new tab instead.

In general we don't have too much option with nomadhelper.html, because from Royale perspective app it's just a page/html file which is being opened inside app in iFrame. I need to investigate whether I can somehow "contact" with opened in iFrame nomadhelper.html or not. @Aszusz mention about some API, which I may try soon. This issue would be the last one to complete this design fully.

**Comment by JoelProminic** [comment]
Some ideas for communicating with the iframe to detect errors:
https://github.com/ternarylabs/porthole
https://stackoverflow.com/questions/9153445/how-to-communicate-between-iframe-and-the-parent-site

However, I think we can do some of this inside nomadhelper.html
- For the [else case](https://github.com/Moonshine-IDE/Super.Human.Portal/blob/ff996d425e4ba36ef01549512d9f580d4e3fbcb7/Super.Human.Portal_Royale/src/resources/nomadhelper.html#L31), open the URL ina new tab instead of just logging a message
- Check the results of the [postMessage](https://github.com/Moonshine-IDE/Super.Human.Portal/blob/ff996d425e4ba36ef01549512d9f580d4e3fbcb7/Super.Human.Portal_Royale/src/resources/nomadhelper.html#L20-L28) and open the URL in a new tab if it fails.

We'll also need to handle the case where nomadhelper.html doesn't exist - this indicates that we made a mistake while configuring the Vagrant instance, but it shoudln't need to block this action.

**Comment by piotrzarzycki21** [comment]
@JoelProminic I was able to sent message from nomadhelper to our application. I'm displaying message to the user which informs that nomad is not opened - message can be change to something more extensive. Changes are in newest build. You have to deploy new version of nomadhelper.html.

**Comment by JoelProminic** [comment]
As discussed in the meeting today, Super.Human.Portal or nomadhelper.html should open the Nomad link in a new tab if there was any error in nomadhelper.html.

Also, show a message on the bottom of the screen when the link is opened (even if nomadhelper.html worked properly) so that the user knows that they should check the Nomad tab if it didn't focus automatically.

**Comment by piotrzarzycki21** [comment]
> As discussed in the meeting today, Super.Human.Portal or nomadhelper.html should open the Nomad link in a new tab if there was any error in nomadhelper.html.
> 
> Also, show a message on the bottom of the screen when the link is opened (even if nomadhelper.html worked properly) so that the user knows that they should check the Nomad tab if it didn't focus automatically.

@JoelProminic I have pushed changes which opens new tab if Service worker is not registered. 
There is unfortunately one more case which is completely out of my hands. Here is scenario:

1) Open Nomad Web
2) Open link from Super.Human.Portal - Works
3) Close Nomad Web Tab
4) Try to Open link from Super.Human.Portal - Bug - service worker still exists, so SHP reports that application is opened in Nomad Web tab

**Comment by JoelProminic** [comment]
@piotrzarzycki21, I spent a little time checking if [serviceWorker.controller.postMessage](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorker/postMessage) reports an error somehow.  

The most promising option I found was the [error event on the serviceWorker.controller object](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorker/error_event).

It looks like our best bet would be to check the [message event on the serviceWorker object](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerContainer/message_event) event, but these won't necessarily all be errors.  I do see an [error event](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerContainer#events) as well, but it is deprecated.

**Comment by JoelProminic** [comment]
Testing with today's updates, I nomadhelper.html is not working at all for me.  Even if Nomad is open, the database is not opened in the Nomad tab.    Testing with the debugger, I see that the snackbar message is sent before the postMessage is even called.

**Comment by piotrzarzycki21** [comment]
@JoelProminic I have added to nomadhelper.html following code. Unfortunately when I compared objects which I'm getting in events between Opened vs Closed Nomad tab - there is no clue inside that objects whether Nomad Tab is opened. 
Events objects looks the same in both cases. Try yourself by adding that code and deploying it in newest nomadhelper.html.

```
f (navigator && navigator.serviceWorker && navigator.serviceWorker.controller) {
        		navigator.serviceWorker.oncontrollerchange = function(event) {
        			console.log('ServiceWorker: [oncontrollerchange]');
        	    		console.log(event);
        	    }
        	    navigator.serviceWorker.onmessage = function(event) {
        	    		console.log('ServiceWorker: [onmessage]');
        	    		console.log(event);
        	    }
        	     navigator.serviceWorker.onmessageerror = function(event) {
        	     	console.log('ServiceWorker: [onmessageerror]');
        	    		console.log(event);
        	    }
        	    
        	    navigator.serviceWorker.controller.onstatechange = function(event) {
        	     	console.log('Controller: [onstatechange]');
        	    		console.log(event);
        	    }
        	    
        	    navigator.serviceWorker.controller.onerror = function(event) {
        	     	console.log('Controller: [onerror]');
        	    		console.log(event);
        	    }
        	    
            navigator.serviceWorker.controller.postMessage({
                type: 'openNotesUri',
                payload: {
                    notesUri: decodeURIComponent(
                        hrefUrl.hash.substring(
                            hrefUrl.hash.indexOf(HASH_PREFIX) + HASH_PREFIX.length
                        )
                    )
                }
            });
            console.log(navigator.serviceWorker);
            console.log(navigator.serviceWorker.controller);
            window.top.postMessage('[Success] Successfully called openNotesUri', '*');
        } else {
            console.error('No service worker registered');
                    
        		window.top.postMessage('[Error] No service worker registered', '*');
        }
```

**Comment by piotrzarzycki21** [comment]
@JoelProminic branch has been merged to main.

**Comment by JoelProminic** [comment]
Suggestions from Daniel from HCL

>The scopedlinks.html page is itself a client to the Nomad service worker.  If itself and Nomad are opened, you'll see the two clients listed in Chrome Dev Tools:
> 
>If Nomad was closed, but scopedlinks.html is still open, just the one client is listed:
> 
>Any checks that that page performs against navigator.serviceWorker.controller will be a non-null value, assuming the service worker is installed.
> 
>I believe what could help with your scenario is instead of checking navigator.serviceWorker.controller, you can check the client count. This is queried from the service worker as it's a service-worker only API ( https://developer.mozilla.org/en-US/docs/Web/API/Clients ).  If the Nomad service worker finds more than one window client opened, then we can assume that a Nomad window is also opened, and you can continue with the postMessage to open the link.
> 
>First you'll need to register a service worker message listener in scopedlinks.html ( https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerContainer/message_event ). Listen for the event.data.type 'numberOfWindowClients'.  When the service worker posts a message payload back to the client, you can get the number of window clients in this message handler via event.data.clientCount.
> 
>Then, to query the service worker for the client count, perform a postMessage to the service worker ( https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorker/postMessage ) with message { type: 'getNumberOfWindowClients' }
> 
>Note that this is an asynchronous operation because it's message-driven. You'll have to wait for the client count response before moving forward with the rest of the logic.

**Comment by piotrzarzycki21** [comment]
> Then, to query the service worker for the client count, perform a postMessage to the service worker ( https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorker/postMessage ) with message { type: 'getNumberOfWindowClients' }

Unfortunately whatever I tried this part doesn't work. I will share details during our today's meeting.

**Comment by JoelProminic** [comment]
From our testing today, it looks like the `postMessage` should be `numberOfWindowClients` instead of `getNumberOfWindowClients`.

The flow would be:
1. Super.Human.Portal calls nomadhelper.html with a URL
2. nomadhelper.html sets up the message handlers
3. nomadhelper.html sends a `postMessage` with `numberOfWindowClients`
4. `message` event triggers with `type:numberOfWindowClients`
5. Check the count
    1. If it indicates that there is an active client, submit the openNotesUri postMessage as before
    2. If it indicates that there is not an active client, report an error back to Super.Human.Portal

**Comment by JoelProminic** [comment]
I did some digging in the service worker from the Firefox debug console, and I parsed a list of likely commands from `/nomad/sw-bundle.js`.   If you see something that may be useful, you can dig through the JavaScript to try and figure out what it does, but the code is minimized.

    authenticate
    authenticated
    broadcastHideBackdrop
    broadcastNomadUpdating
    broadcastResetNomad
    broadcastShowBackdrop
    claimClients
    clientClosing
    createIFrame
    end
    focusPrimaryWindow
    fromWaitingSW-returnNewAndCurrentVersion
    getSWVersion
    newClient
    numberOfWindowClients
    openNotesUri
    skipWaiting
    startDownload
    toWaitingSW-returnNewAndCurrentVersion
    updateOnlineOffline

**Comment by JoelProminic** [comment]
Testing the latest nomadhelper.html and Super.Human.Portal on the Prominic servers, I noticed that nomadhelper.html seems to keep triggering the URLs.  I saw the database open multiple times, and they would reopn after I closed them.   The Prominic Nomad server was running Nomad "1.0.10.5955-3678"

I'll do some more tests to see if this triggers on SHI servers as well.

**Comment by piotrzarzycki21** [comment]
> * Close Nomad and reload Super.Human.Portal
> * Click an Open in Nomad button.  BUG:  This is getting blocked as a popup with the latest build - this wasn't triggering in my other recent tests.  I allowed the individual popup to continue.   See example below

@JoelProminic I have recorded video - it looks like on my sight it doesn't trigger or I don't understand what do you mean by that. 


https://github.com/Moonshine-IDE/Super.Human.Portal/assets/24554795/3a813068-30b5-4d60-9f11-7894bd40c4d2

**Comment by piotrzarzycki21** [comment]
> 7\. BUG:  The database reopens after a short delay.

This part I have reproduced. In newest build it's fixed.

**Comment by JoelProminic** [comment]
I'm still getting the popups on domino-49.prominic.net.   Check your popup settings and exclusions to see if you excluded the test server previously.  For Firefox, this is FireFox > Settings > Privacy & Security > Block pop-up windows > Exceptions.

**Comment by piotrzarzycki21** [comment]
What about Bug number 7?

**Comment by JoelProminic** [comment]
In addition to the popup bar, I get this message on the console:

    Uncaught Error: navigateToURL requires interaction trigger
        navigateToURL https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:41083
        controller_CommandLaunchNomadLink_onWindowMessage https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:37030
        execute https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:37001
        executeCommand https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:53191
        notifyObserver https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:36861
        notifyObservers https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:19369
        notifyObservers https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:43936
        sendNotification https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:43932
        sendNotification https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:12519
        mediator_applications_MediatorInstalledApps_onOpenInNomadLink https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:42557
        fireListeners https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:12857
        dispatchEventInternal_ https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:9429
        dispatchEvent https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:12837
        dispatchEvent https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:12984
        view_controls_LinkWithDescriptionAppButton_onLinkClick https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:55878
        $EH_13_0 https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:55903
        fireListener https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:9215
        fireListenerOverride https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:12903
        handleBrowserEvent_ https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:9259
        f https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:9021
        addEventListener eval:4
        listen_ https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:9007
        listen https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:8970
        addEventListener https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:12948
        initializeStrandBasedObject https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:57650
        generateMXMLArray https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:57552
        generateMXMLInstances https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:57676
        addedToParent https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:14510
        addElement https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:13597
        addElement https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:14395
        initializeStrandBasedObject https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:57665
        generateMXMLArray https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:57552
        generateMXMLInstances https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:57676
        addedToParent https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:14510
        addElement https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:13597
        addElement https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:14395
        mediator_applications_MediatorInstalledApps_updateInstalledAppLinks https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:42503
        mediator_applications_MediatorInstalledApps_updateView https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:42458
        onRegister https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:42433
        registerMediator https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:19398
        registerMediator https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:43913
        execute https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:22738
        executeCommand https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:53191
        notifyObserver https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:36861
        notifyObservers https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:19369
        notifyObservers https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:43936
        sendNotification https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:43932
        sendNotification https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:12519
        mediator_MediatorMainContentView_onNavigationInstalledAppSectionChange https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:36383
        fireListener https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:9215
    SuperHumanPortal_Royale.js:41083:11



The database reopening issue is resolved in my quick test.  I'll report if I start seeing it again.

**Comment by piotrzarzycki21** [comment]
> In addition to the popup bar, I get this message on the console:
> 
> ```
> Uncaught Error: navigateToURL requires interaction trigger
>     navigateToURL https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:41083
>     controller_CommandLaunchNomadLink_onWindowMessage https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:37030
>     execute https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:37001
>     executeCommand https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:53191
>     notifyObserver https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:36861
>     notifyObservers https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:19369
>     notifyObservers https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:43936
>     sendNotification https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:43932
>     sendNotification https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:12519
>     mediator_applications_MediatorInstalledApps_onOpenInNomadLink https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:42557
>     fireListeners https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:12857
>     dispatchEventInternal_ https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:9429
>     dispatchEvent https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:12837
>     dispatchEvent https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:12984
>     view_controls_LinkWithDescriptionAppButton_onLinkClick https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:55878
>     $EH_13_0 https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:55903
>     fireListener https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:9215
>     fireListenerOverride https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:12903
>     handleBrowserEvent_ https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:9259
>     f https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:9021
>     addEventListener eval:4
>     listen_ https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:9007
>     listen https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:8970
>     addEventListener https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:12948
>     initializeStrandBasedObject https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:57650
>     generateMXMLArray https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:57552
>     generateMXMLInstances https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:57676
>     addedToParent https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:14510
>     addElement https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:13597
>     addElement https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:14395
>     initializeStrandBasedObject https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:57665
>     generateMXMLArray https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:57552
>     generateMXMLInstances https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:57676
>     addedToParent https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:14510
>     addElement https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:13597
>     addElement https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:14395
>     mediator_applications_MediatorInstalledApps_updateInstalledAppLinks https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:42503
>     mediator_applications_MediatorInstalledApps_updateView https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:42458
>     onRegister https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:42433
>     registerMediator https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:19398
>     registerMediator https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:43913
>     execute https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:22738
>     executeCommand https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:53191
>     notifyObserver https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:36861
>     notifyObservers https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:19369
>     notifyObservers https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:43936
>     sendNotification https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:43932
>     sendNotification https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:12519
>     mediator_MediatorMainContentView_onNavigationInstalledAppSectionChange https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:36383
>     fireListener https://<-snip->/Super.Human.Portal/js-debug/SuperHumanPortal_Royale.js:9215
> SuperHumanPortal_Royale.js:41083:11
> ```
> 
> The database reopening issue is resolved in my quick test. I'll report if I start seeing it again.
> 
> UPDATE: I'm not sure what you are looking for with "Bug number". This is the same issue I reported for step 2 in my last test procedure:
> 
> > 2. Click an Open in Nomad button. BUG: This is getting blocked as a popup with the latest build - this wasn't triggering in my other recent tests. I allowed the individual popup to continue. See example below

I have cleaned up above exception - it occurs when browser blocking popups window.open returns null in that case and internal logic of [navigateToURL throwing exeption](https://github.com/apache/royale-asjs/blob/4039fec9df3841652a950d30b96de699faa9d0c3/frameworks/projects/Network/src/main/royale/org/apache/royale/net/navigateToURL.as#L70).

**Comment by piotrzarzycki21** [comment]
> 2\. Click an Open in Nomad button.  BUG:  This is getting blocked as a popup with the latest build - this wasn't triggering in my other recent tests.  I allowed the individual popup to continue.   See example below

@JoelProminic I'm not sure what is the expectation here. I'm able to reproduce scenario where browser is blocking my popup by removing it from Block pop-up windows > Exceptions. However I don't think it is my responsibility to do something more for the user. I do see space for displaying Snackbar message when popup is blocked because navigateToURL throwing exception. Do you want me to do this ?

**Comment by JoelProminic** [comment]
@piotrzarzycki21, I wasn't seeing the Nomad links blocked by the popup blocker until the changes this week, so I'd like to approach this by figuring out what has changed.  Does this happen because we are using `navigateToURL` different, or trying to open the links from nomadhelper.html?

I feel the (blocked) popups are going to create a  bad user experience.  Even if this will be primarily be used by the Prominic support team, there are some users who will immediately lose respect for the application if they have too mess with popups.

If you don't see any good leads for this, we can discuss it more in the meeting on Tuesday.

**Comment by JoelProminic** [comment]
To clarify:
1. Clear nomad_helper_url in SuperHumanPortal.nsf
2. Refresh Super.Human.Portal
3. Click a button to trigger the Open in Nomad logic
4. Opens a new tab.  This is not blocked by the popup blocker
5. Set nomad_helper_url
6. Close Nomad tab
7. Refresh Super.Human.Portal
8. Click a button to trigger the Open in Nomad logic
9. BUG:  triggers popup blocker
10.  Allow this popup.  It opens a new tab which launches Nomad.

**Comment by piotrzarzycki21** [comment]
I believe there is some ideas how to avoid blockers [here](https://stackoverflow.com/questions/2587677/avoid-browser-popup-blockers). I just need to experiment to do this. I'm not sure if this is suppose to be a blocker.

**Comment by JoelProminic** [comment]
The first answer to that question is my understanding of the browser popup-blocker behavior.   The browser is not able to trace the button click through to the window open in this case, so it blocks the action.

The second action looks promising, though I'm not clear on how it will look to the user.  If it is opening and closing the new tab and changing the user focus, that could be distracting. 

I feel that this is a blocker, since it seems like the kind of behavior that support would need to guide non-technical users through, and that technical  users would find suspicious..   I remember @JustinProminic also had concerns for this, but he may decide to bump this to the next release if he wants to get this release fully pushed out.

**Comment by JoelProminic** [comment]
Here is a new issue I found while testing on the Prominic server.  nomadhelper.html got an error message indicating that the Nomad window had closed, but it did not attempt to open a new tab:

    Nomad Window is closed nomadhelper.html:39:15
        messageFunction https://-snip-:9443/nomad/nomadhelper.html?link=https://-snip-:9443/nomad/#/notes://-snip-?OpenView:39
        (Async: EventListener.handleEvent)
        openLink https://-snip-:9443/nomad/nomadhelper.html?link=https://-snip-:9443/nomad/#/notes://-snip-?OpenView:47
        onload https://-snip-:9443/nomad/nomadhelper.html?link=https://-snip-:9443/nomad/#/notes://-snip-?OpenView:12
        (Async: EventHandlerNonNull)
        <anonymous> https://-snip-:9443/nomad/nomadhelper.html?link=https://-snip-:9443/nomad/#/notes://-snip-?OpenView:9

**Comment by JoelProminic** [comment]
@JustinProminic did tests with today, and noticed these problems:
- When clicking an unexpanded button, the Nomad tab does not open
    1. Close any Nomad tabs
    2. Open Super.Human.Portal and login
    3. Bookmarks/Default
    4. Create a database bookmark if none is present (For example SuperHumanPortal.nsf on your server)
    5. While the bookmark is collapsed, click the bookmark. BUG:  Nothing happens.  This is expected to trigger the Open in Nomad logic by default
    6. Expand the bookmark and click Open in Nomad.  This works.
- He hit the popup blocker, and found that the notification was not obvious.
- The databases were not opening when the Nomad tab was open.  Justin reopened the tab, and that fixed the problem.  I have noticed some promblems like this, usually the databases eventually open after a delay.

I confirmed that nomadhelper.html was up-to-date on the Nomad server.

Justin suggested that we show a message telling the user how to get past the popups.  Ideally this would have arrows showing users where to click.

@MarkProminic found [this option](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/browserSettings/allowPopupsForUserEvents) for changing the browser popup behavior. This has limited browser support, though.

**Comment by JoelProminic** [comment]
Unfortunately, it looks to me like `browserSettings.allowPopupsForUserEvents` is a feature for browser extensions rather than for general JavaScript.  This also looks to be a global setting, so it wouldn't make sense to allow unrestricted access to websites.

I think the solution Piotr found is the best option for resolving the popup blocker events.   If this doesn't work, then we may need to switch over to @JustinProminic's design.

**Comment by piotrzarzycki21** [comment]
> 5\. While the bookmark is collapsed, click the bookmark. BUG:  Nothing happens.  This is expected to trigger the Open in Nomad logic by default

@JoelProminic I have fixed this bug. It's in newest build.

**Comment by piotrzarzycki21** [comment]
> The first answer to that question is my understanding of the browser popup-blocker behavior. The browser is not able to trace the button click through to the window open in this case, so it blocks the action.
> 
> The second action looks promising, though I'm not clear on how it will look to the user. If it is opening and closing the new tab and changing the user focus, that could be distracting.
> 
> I feel that this is a blocker, since it seems like the kind of behavior that support would need to guide non-technical users through, and that technical users would find suspicious.. I remember @JustinProminic also had concerns for this, but he may decide to bump this to the next release if he wants to get this release fully pushed out.

@JoelProminic proposed solution with opening empty window and replace url from SOW didn't work for me. I've decided to inform use what to do when popup is blocked. I'm displaying following message. We should discuss if this is acceptable solution tomorrow. My changes are pushed so you can test it. 

<img width="1792" alt="Screenshot 2024-05-27 at 20 20 27" src="https://github.com/Moonshine-IDE/Super.Human.Portal/assets/24554795/fe9ec541-c600-4432-a07d-8e4fbc738b53">

**Comment by piotrzarzycki21** [comment]
@JoelProminic I think I have fixed this NPE which we have experienced during meeting. Newest build should have my changes.

**Comment by JoelProminic** [comment]
I confirmed that the issue when switching categories under Bookmarks was resolved.

There is a remaining issue where a database opens twice when the Nomad window was not open, but this seems to be an issue on the Nomad side.
